### ISSUE TITLE
Crafting roadmap: UI improvements - collapsible tool/ingredient details

### ISSUE THREAD

**Issue by I-am-Erk** [CleverRaven/Cataclysm-DDA#62634]
### Is your feature request related to a problem? Please describe.

We have a lot of changes to look at for our crafting system, but these come in the context of a crafting UI that is already extremely bulky and hard to parse, for new and old users alike. It *needs updating*, one way or another, and the question is just one of how to update it. As we fix things with crafting and allow stuff like variable time and tool choices, we are going to have exponentially more and more information to display on an already cluttered window.

In resolving this issue I believe fairly strongly that we should be reducing base complexity. Most of our crafting window information is redundant and unnecessary 90% of the time. It should be accessible when needed, but it doesn't need to be accessible at all times.

![image](https://user-images.githubusercontent.com/45136638/206568818-b76e876d-77d9-4d00-a839-ca41c4a2adb5.png)

For this issue, I will prototypically look at that big, ugly, almost always useless block of red letters and `OR`.

### Solution you would like.

1. Add a mandatory `"name"` attribute to `requirement` JSON types. For example:

```

  {
    "id": "surface_heat",
    "type": "requirement",
    "name": "heat source",
    "//": "Heat usable for heating a surface - for example a pot or frying pan.",
    "tools": [
      [
        [ "hotplate", 1 ],
```
2. If there are more than two options available for a particular recipe tool or ingredient, collapse them and instead display the requirement name. For example:  `heat source (8 charges)`. It may be necessary to show a charge range, eg "8-12 charges". Fire should be a special case, so in the above example we'd see `heat source (8 charges) or nearby fire`.
    - If requirements are partly met, we leave the field collapsed. So if you have a heat source with 3 charges, the entirety of "heat source (8 charges)" will appear in brown.
    - If requirements are met, show the met requirements first, in green. So if you have an oven and a fire nearby you'd see `"Nearby fire or oven (8 charges)"`. I am not sure if we need to comment here on there being other heat sources available, it's possibly reasonable to follow that in red with `"or other heat source (8 charges)"`. As above, this collapses if there are more than two options, so if you have an oven and a hot plate and a nearby fire you'd see `"heat source (8 charges) or nearby fire"` in green.

3. Add a "show details" toggle key. When toggled, the item information pane is replaced with a full list of all tools and ingredients required for the recipe, using the same format we now have. I suggest we put this with "tools required" so nobody misses it, eg:
```
Tools required (press [R] for details):
```
The same applies to components required and they can probably use the same details pane and hotkey.

---
PR these changes at this point. What follows should be a second PR or series of PRs depending on how much churn it generates.
---

4. Add a code CI check that looks for recipes that offer more than two options for an ingredient, and throw a JSON error. These should be replaced with requirement lists.
5. Go through all the recipes and add requirement lists for everything that has more than two options for an ingredient.

### Describe alternatives you have considered.

I genuinely don't have another way here. These lists have become impossibly huge and will only grow, and players only rarely need this information. Hiding it into a key toggle will almost never cause extra work, and when it does, it's at a time when the extra work is ok.

The specifics of how we show and implement this could vary. Particularly, I am not entirely sure how we should show the collapsed information.

### Additional context

We still need to collapse some of the other information in this window in a similar way. I also want to devise a basic algorithm to automatically select tools and ingredients, such that you only need to manually select them if you want to be particular about what you use to craft.

**Comment by PatrikLundell** [comment]
I fail to see any benefit to replacing large and cumbersome lists of all the alternatives with vague indications of what the lists might contain plus unused space. I understand that micro display users may want to compress things such that they can see all of the vague indications without scrolling, but I don't think penalizing users of real screens is the way to do it. If a toggle was to be introduced, I would suggest the introduction of a game option specifying the default state of the toggle, so micro screen users (and others) who'd rather use hints and no scrolling can have that as the default, while the rest default to display it all.

4. and 5. should be reversed in order. Don't introduce a massive reject everything shit storm because master doesn't live up to new requirements. Instead, fix master and only then introduce rejections. If it's possible to selectively reject new "violations" only, without triggering on preexisting violations in master, that would be reasonable to do before master has been processed fully, but not before master conversion has been started.

I'm not that keen on auto selection of ingredients, as that inevitably will lead to players losing scarce (for them) resources inadvertently because they didn't think to double check what the auto selection decided upon. Not every time, but often enough to be a problem.

**Comment by Zireael07** [comment]
IMHO named requirements are a good idea, but a "show details" toggle is NOT. People will forget to press it and come complaining that they can't find things/don't know things (remember not everyone is a veteran player :P)

Seconding the idea that it could be useful for micro screen users, though.

**Comment by Raikiri** [comment]
don't we already have a mechanism to describe a wide range of tools? it's called qualities. and most tools do follow it: if a recipe requires a tool with hammering 1, we don't have to list all tools that have this quality. why do we have to list tools that heat things up then?

**Comment by GuardianDll** [comment]
Resolved by #86044 and it's predecessors i would assume
